instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java

```diff
@@ -54,189 +42,63 @@
-
-  private final boolean hystrixTagsEnabled;
-
-  private final Map<String, String> endpointPeerServiceMapping;
-
-  // Values from an optionally provided properties file
-  private static Properties propertiesFromConfigFile;
-
-  // Read order: System Properties -> Env Variables, [-> properties file], [-> default value]
-  // Visible for testing
-  Config() {
-    propertiesFromConfigFile = loadConfigurationFile();
-
-    propagators = getListSettingFromEnvironment(PROPAGATORS, null);
-    exporterJar = getSettingFromEnvironment(EXPORTER_JAR, null);
-    exporter = getSettingFromEnvironment(EXPORTER, "otlp");
-    traceEnabled = getBooleanSettingFromEnvironment(TRACE_ENABLED, DEFAULT_TRACE_ENABLED);
-    integrationsEnabled =
-        getBooleanSettingFromEnvironment(INTEGRATIONS_ENABLED, DEFAULT_INTEGRATIONS_ENABLED);
-
-    excludedClasses = getListSettingFromEnvironment(TRACE_CLASSES_EXCLUDE, null);
-
-    scopeDepthLimit =
-        getIntegerSettingFromEnvironment(SCOPE_DEPTH_LIMIT, DEFAULT_SCOPE_DEPTH_LIMIT);
-
-    runtimeContextFieldInjection =
-        getBooleanSettingFromEnvironment(
-            RUNTIME_CONTEXT_FIELD_INJECTION, DEFAULT_RUNTIME_CONTEXT_FIELD_INJECTION);
-
-    traceAnnotations = getSettingFromEnvironment(TRACE_ANNOTATIONS, DEFAULT_TRACE_ANNOTATIONS);
-
-    traceMethods = getSettingFromEnvironment(TRACE_METHODS, DEFAULT_TRACE_METHODS);
-    traceAnnotatedMethodsExclude =
-        getSettingFromEnvironment(
-            TRACE_ANNOTATED_METHODS_EXCLUDE, DEFAULT_TRACE_ANNOTATED_METHODS_EXCLUDE);
-
-    traceExecutorsAll =
-        getBooleanSettingFromEnvironment(TRACE_EXECUTORS_ALL, DEFAULT_TRACE_EXECUTORS_ALL);
+  private static final Config DEFAULT = Config.newBuilder().build();
 
-    traceExecutors = getListSettingFromEnvironment(TRACE_EXECUTORS, DEFAULT_TRACE_EXECUTORS);
+  // INSTANCE can never be null - muzzle instantiates instrumenters when it generates
+  // getInstrumentationMuzzle() and the Instrumenter.Default constructor uses Config
+  private static volatile Config INSTANCE = DEFAULT;
 
-    sqlNormalizerEnabled =
-        getBooleanSettingFromEnvironment(SQL_NORMALIZER_ENABLED, DEFAULT_SQL_NORMALIZER_ENABLED);
-
-    kafkaClientPropagationEnabled =
-        getBooleanSettingFromEnvironment(
-            KAFKA_CLIENT_PROPAGATION_ENABLED, DEFAULT_KAFKA_CLIENT_PROPAGATION_ENABLED);
-
-    hystrixTagsEnabled =
-        getBooleanSettingFromEnvironment(HYSTRIX_TAGS_ENABLED, DEFAULT_HYSTRIX_TAGS_ENABLED);
-
-    endpointPeerServiceMapping = getMapSettingFromEnvironment(ENDPOINT_PEER_SERVICE_MAPPING);
-
-    log.debug("New instance: {}", this);
+  /**
+   * Sets the agent configuration singleton. This method is only supposed to be called once, from
+   * the agent classloader just before the first instrumentation is loaded (and before {@link
+   * Config#get()} is used for the first time).
+   */
+  public static void internalInitializeConfig(Config config) {
```

Comment:
I usually prefer method names like `doSomethingInternal` vs `internalDoSomething`. What do others think?

Reply:
I have slight preference for `internalDoSomething` - gets the point across that it's internal sooner. Another option could be to make the access `package-private` and move the method to an `InternalConfigAccess` type of class.

Reply:
interesting, I've typically done `doSomethingInternal`, because it is often a delegate of `doSomething`, but I like the reasoning of putting internal in front to "get the point across that it's internal sooner"